### PR TITLE
Merge castellanos70 1f2e49a with kt stephano 93b7add

### PR DESCRIPTION
**Pull request by max-sanchez**
This commit makes fairly major changes to the comm module.

All files in the AITest have been deleted, which is great because they were out-of-sync with the actual AI development.

Justin Hall's comments:
The biggest update is to the ConcurrentCommModule which was partially rewritten.

Major Changes:
- Comm module can now receive data from the server at the same time it is polling previously-sent data
- Server.java now contains 3 constructors: 2 public, 1 private (more to be added later)
- PackedTileData now implements Sendable so it can be sent by the server over the network
- When the game is first starting, the server now sends the PackedTileData built by the simulator during its initialization
- Server.jar has been updated to the latest version of the server

Known Bugs:
- Comm module sometimes has trouble closing the program when Server.jar was not found. System.exit never seems to complete in these cases.

### CODE REVIEW

**Comment by kf4x** [comment]
@Vorksholk have you tested this?

**Comment by kf4x** [comment]
@KTStephano what client class do i run to test this?

**Comment by max-sanchez** [comment]
@javierchavez I compiled and ran the server, 2xAI, and UpdateLoop client with assertions enabled.

Out of 6 runs, 4 of them ran successfully (which is about average), with one of the failures occurring with the AI card hand=0 issue, and another where the GUI doesn't display the cards.

Both of those issues appear related to the shared networking/client code not properly receiving the hand from the server at initial startup, and have existed for quite a while in the code. 

I also ran all of the unit tests which appear as main methods:
starvationevasion.common.MapPoint
starvationevasion.common.SeaLevel
starvationevasion.sim.Simulator (which displays plausible world data)
starvationevasion.sim.CardDeck
starvationevasion.PreprocessingTools.CoordinateTesting (reports "Start viable")

You mentioned a bunch of unit tests when talking about the recent policy card pull request. If you send those to me, I'd be happy to run those as well when I do my testing.

**Comment by KTStephano** [comment]
@javierchavez You can either start the Server and then UpdateLoop/AI after it's running, or you can start the UpdateLoop/the AI and they will spawn their own server and connect to it.

**Comment by kf4x** [comment]
@KTStephano it didn't do anything. I wanted to connect to foodgame server so I changed the host in `UpdateLoop` and nothing seems to come up.

``` java
client = new ClientTest(this, "foodgame.cs.unm.edu", 5555);
```

**Comment by kf4x** [comment]
@Vorksholk I havent had a chance to change anything today with regard to the tests. are you using windows or unix machine

**Comment by KTStephano** [comment]
@javierchavez Hmm I'm not sure what's causing that. I just ran a test on mine and it connected to foodgame pretty fast and the game started almost immediately. Does the GUI come up at least?

**Comment by max-sanchez** [comment]
@Javier Chavez Windows 7 x64, Java 8u97, all testing done with a local server.

**Comment by kf4x** [comment]
@Vorksholk Crap you will need Unix linux or mac, my tests are all written in bash and use netcat. 
Thats why I rely so heavily on `Game` file.

**Comment by max-sanchez** [comment]
@javierchavez No problem, I can run a VM, or throw it on a headless server. I tried running your Game script a few weeks ago on headless linux w/ Java 8uXX on it and had a few issues, I'll look into it later today and I might have some questions.

**Comment by kf4x** [comment]
@KTStephano No gui. [This](https://github.com/castellanos70/StarvationEvasion/blob/master/src/starvationevasion/client/UpdateLoop.java#L140) is exactly the line I changed. then commented out the next one

**Comment by kf4x** [comment]
@KTStephano I get:

```
Client: Attempting to connect to host foodgame.cs.unm.edu, port 5555
Client: Connection successful
```

But no GUI

**Comment by KTStephano** [comment]
@javierchavez Oh wow that's really weird. So the only time it prints that is if it's connected and performed the full handshake with the server without any problems coming up. Usually the GUI pops up right after that though.

**Comment by kf4x** [comment]
@KTStephano I need to go to sleep, otherwise I would test out your fork to see if there is an issue with newer code affecting it.
I checked your repo and saw

```
This branch is 9 commits ahead, 70 commits behind castellanos70:master
```

**Comment by kf4x** [comment]
@Vorksholk would you mind trying as well. 
Uncomment this [line](https://github.com/castellanos70/StarvationEvasion/blob/master/src/starvationevasion/client/UpdateLoop.java#L140) and comment out the next one.

**Comment by KTStephano** [comment]
@javierchavez Let me update my repo and run the test again.

**Comment by max-sanchez** [comment]
@javierchavez I tried connecting the GUI to foodgame.cs.unm.edu:5555, and it fails. Is foodgame running the latest version of the server from this commit, though? I would assume changes by @KTStephano and co. would necessitate an update on foodgame because of the changes to the communication module.

**Comment by kf4x** [comment]
no it doesnt matter. from what I read, they didn't change anything that effects how the server actually handles connections. But last night I went ahead and pulled the code onto the foodgame machine.

TL;DR yes foodgame has this merge. glad we are consistent with error.

**Comment by KTStephano** [comment]
@javierchavez Hmm I still can't reproduce the error after updating to the latest code. I'm not too sure why it doesn't consistently work.
Here is the output I get after commenting out the localhost line (after logging in through the GUI):

```
Client: Attempting to connect to host foodgame.cs.unm.edu, port 5555
Client: Connection successful
Client: Login successful
Starting game . . .
```

**Comment by KTStephano** [comment]
@javierchavez @Vorksholk Would you guys mind either running in debug mode with a break just after line 40 of starvationevasion.client.Networking.ClientTest.java or maybe put a print statement on line 41? I really want to know if our code is actually returning from the connect sequence so that the rest of the client code gets to execute.

**Comment by max-sanchez** [comment]
@javierchavez @KTStephano I just tested the UpdateLoop again with the connection to foodgame.cs.unm.edu with KTStephano's changes, and everything worked fine (could get a hand, draft cards, etc.). 

Maybe FoodGame was full/down/otherwise unavailable earlier when I tried?

**Comment by kf4x** [comment]
@Vorksholk I haven't changed a thing. Like I said I pulled & started it last night and havent touched it since.

**Comment by javierchavez** [comment]
@Vorksholk @KTStephano I am testing again right now

**Comment by javierchavez** [comment]
@Vorksholk @KTStephano anyone at UNM

**Comment by kf4x** [comment]
@Vorksholk @KTStephano Ill be working with this more over summer so ill to clean up and get rid of all these hacks and brittleness that @KTStephano added 😃. but in seriousness, its important for you all to be able to make progress and sometimes hacks are what you gotta do until you have more time. its your call @Vorksholk, if you think this is ok and stable then merge it.

**Comment by max-sanchez** [comment]
@javierchavez Alright, I'm gonna merge this in, cause I gotta merge in a backlog of other items. Thanks for your work.
